pkg/upgrade/upgrade.go

```diff
@@ -257,175 +283,213 @@ func UpdateFromLegacyVersion(cli client.Client, platform deploy.Platform, appNS
+				Values:    []string{"rhods-model-monitoring"},
+			},
+			{
+				Gvk:       gvkSecret,
+				Namespace: dscMonitoringNamespace,
+				Path:      pathName,
+				Values:    []string{"rhods-monitoring-oauth-config"},
+			},
+			{
+				Gvk:       gvkClusterRole,
+				Namespace: dscMonitoringNamespace,
+				Path:      pathName,
+				Values:    []string{"rhods-namespace-read", "rhods-prometheus-operator"},
+			},
+			{
+				Gvk:       gvkClusterRoleBinding,
+				Namespace: dscMonitoringNamespace,
+				Path:      pathName,
+				Values:    []string{"rhods-namespace-read", "rhods-prometheus-operator"},
+			},
+			{
+				Gvk:       gvkServiceAccount,
+				Namespace: dscMonitoringNamespace,
+				Path:      pathName,
+				Values:    []string{"rhods-prometheus-operator"},
+			},
+			{
+				Gvk:       gvkServiceMonitor,
+				Namespace: dscMonitoringNamespace,
+				Path:      pathName,
+				Values:    []string{"modelmesh-federated-metrics"},
+			},
+		}
+		toDelete = append(toDelete, del...)
+	}
 
-		deprecatedSecrets := []string{"rhods-monitoring-oauth-config"}
-		multiErr = multierror.Append(multiErr, deleteDeprecatedResources(ctx, cli, dscMonitoringNamespace, deprecatedSecrets, &corev1.SecretList{}))
+	// common logic for both self-managed and managed
+	toDelete = append(toDelete, JobSpec{
+		Gvk:       gvkServiceMonitor,
+		Namespace: dscMonitoringNamespace,
+		Path:      pathName,
+		Values:    []string{"rhods-monitor-federation2"},
+	})
 
-		deprecatedClusterroles := []string{"rhods-namespace-read", "rhods-prometheus-operator"}
-		multiErr = multierror.Append(multiErr, deleteDeprecatedResources(ctx, cli, dscMonitoringNamespace, deprecatedClusterroles, &authv1.ClusterRoleList{}))
+	// Handling for dashboard Jupyterhub CR, see jira #443
+	toDelete = append(toDelete, JobSpec{
+		Gvk:       gvkOdhApplication,
+		Namespace: dscApplicationsNamespace,
+		Path:      pathName,
+		Values:    []string{"jupiterhub"},
+	})
+
+	multiErr = multierror.Append(multiErr, deleteResources(ctx, cli, &toDelete))
+	return multiErr.ErrorOrNil()
+}
 
-		deprecatedClusterrolebindings := []string{"rhods-namespace-read", "rhods-prometheus-operator"}
-		multiErr = multierror.Append(multiErr, deleteDeprecatedResources(ctx, cli, dscMonitoringNamespace, deprecatedClusterrolebindings, &authv1.ClusterRoleBindingList{}))
+func deleteResources(ctx context.Context, c client.Client, jobs *[]JobSpec) error {
+	_, err := deleteResourcesWithDone(ctx, c, jobs)
+	return err
+}
 
-		deprecatedServiceAccounts := []string{"rhods-prometheus-operator"}
-		multiErr = multierror.Append(multiErr, deleteDeprecatedResources(ctx, cli, dscMonitoringNamespace, deprecatedServiceAccounts, &corev1.ServiceAccountList{}))
+// deleteResourcesWithDone takes array of Jobs and deletes all matching resources. Returns true if any was deleted or error.
+func deleteResourcesWithDone(ctx context.Context, c client.Client, jobs *[]JobSpec) (bool, error) {
+	var errors *multierror.Error
+	didWork := false
```

Comment:
I'll just explain my logic behind that, it's definitely place to improve and I agree with your concerns.

So, the original delete*AndCheck checked existence of the resource right after deletion without starting next waiting cycle in ExponentialBackoffWithContext and at the point I did not want to introduce extra delay. And I did not want to add extra check for all the cases to the generic function. So generic function does not know if the deletion actually done or not, but it can signal that it performed some actions (vs there are no objects found for action). That is why the second call if the first one really finished all the work will signal that it did nothing, means for the ExponentialBackoffWithContext that the job is done. 

I think in this place it is not really critical but it changes in theory original logic.

Reply:
I wonder what this extra check gives us. We can essentially say we are done when there's nothing to delete, if there's something we perform delete and retry, as some resources can have finalizers. Not really convinced if that extra immediate `.Get` gives us any benefit.

Maybe an opportunity to simplify here?